CMakeLists.txt

```diff
@@ -14,30 +14,65 @@ if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
   set(GSL_STANDALONE_PROJECT ON)
 endif ()
 
+
+include(CheckCXXCompilerFlag)
+if (NOT MSVC)
+    CHECK_CXX_COMPILER_FLAG("-std=c++14" COMPILER_SUPPORTS_CXX14)
+    CHECK_CXX_COMPILER_FLAG("-std=c++17" COMPILER_SUPPORTS_CXX17)
+    CHECK_CXX_COMPILER_FLAG("-std=c++1z" COMPILER_SUPPORTS_CXX1Z)
+else()
+    CHECK_CXX_COMPILER_FLAG("-std:c++14" COMPILER_SUPPORTS_CXX14)
+    CHECK_CXX_COMPILER_FLAG("-std:c++17" COMPILER_SUPPORTS_CXX17)
+endif()
+
 # when minimum version required is 3.8.0 remove if below
 # both branches do exactly the same thing
-if (CMAKE_VERSION VERSION_LESS 3.7.9)
+#if (CMAKE_VERSION VERSION_LESS 3.7.9)
     if (NOT MSVC)
-        include(CheckCXXCompilerFlag)
-        CHECK_CXX_COMPILER_FLAG("-std=c++14" COMPILER_SUPPORTS_CXX14)
-        if(COMPILER_SUPPORTS_CXX14)
-            target_compile_options(GSL INTERFACE "-std=c++14")
+        if(COMPILER_SUPPORTS_CXX17)
```

Comment:
Just because the compiler supports c++17 doesn't necessarily mean I want to compile my project in that mode. This should be left to the use to decide.

In fact, this library should imho set no flags at all (at most it should `target_compile_features(GSL INTERFACE cxx_std_14)`

The machinery here seems more suited for the tests, but even then, one might want to be able to run multiple tests with the same compiler, but in different language modes.

Reply:
@MikeGitb suggests the right way to do it in modern CMake. CMake 3.8.0 added support for `target_compile_features(GSL INTERFACE cxx_std_`*`)`, which a library should use to suggest the oldest standard they support. From all its calls, CMake calculates which flags to add. Before 3.8.0, I'd guess the correct thing to do is set no `-std=` flag at all and let the top project deal with that.

Reply:
Thanks @MikeGitb, those settings should not affect your project, we only compile tests here. That said, definitely agree with changing the test matrix to support runs with c++14 and c++17 separately, so I am going to revert my changes and explore that other option.
@johelegp consider this an experiment with setting the most c++ standard the compiler supports, for testing the newer features.

Reply:
@annagrin: I admit I haven't tested it, but how does `target_compile_options(GSL INTERFACE "-std=c++17")` not affect my project if I directly or indirectly do `target_link_libraries(my_app GSL)` ?